### PR TITLE
Allow enabling and disabling SSL certificate verification

### LINKED ISSUE #123: Question: Http Test to bypass insecure https url
Hi @fhoeben how can we bypass an insecure https website when using http test fixture?

```bash
|script       |json http test|
| get from | https://xyz.com |
```

I'm getting Caused by: javax.net.ssl.SSLPeerUnverifiedException: Host name 'xyz.com' does not match the certificate subject provided by the peer ...

### PR DESCRIPTION
**Pull request by Fastigium**
This PR is meant to fix #123. It has been tested by hand, and I also performed the manual test mentioned in https://github.com/fhoeben/hsac-fitnesse-fixtures/pull/202#issue-181219021 to verify that this PR does not break enabling/disabling compression.

Automated testing is possible, but would require setting up an entire HttpsServer with a variety of certificates specifically for testing this one command. The test would probably add more code and complexity than the feature itself. Alternatively, badssl.com could be used for automated testing, but [their readme](https://github.com/chromium/badssl.com#disclaimer) specifically says it's meant for manual testing only.

Manual testing can be done using the following test page (requires multiple runs and commenting lines because throwing an exception stops the test):
```
|script  |http test                                    |
|note    |The next 4 commands should throw an exception|
|get from|https://expired.badssl.com/                  |
|get from|https://wrong.host.badssl.com/               |
|get from|https://self-signed.badssl.com/              |
|get from|https://untrusted-root.badssl.com/           |
|disable SSL verification                              |
|note    |The next 4 command should go well (green)    |
|get from|https://expired.badssl.com/                  |
|get from|https://wrong.host.badssl.com/               |
|get from|https://self-signed.badssl.com/              |
|get from|https://untrusted-root.badssl.com/           |
|enable SSL verification                               |
|note    |The next 4 commands should throw an exception|
|get from|https://expired.badssl.com/                  |
|get from|https://wrong.host.badssl.com/               |
|get from|https://self-signed.badssl.com/              |
|get from|https://untrusted-root.badssl.com/           |
```

I hope you'll accept the PR without an automated test, or know of a better way to implement one. Comment/remarks welcome. Full disclosure: this PR is the result of me hunting the $50 bounty offered on [BountySource](https://www.bountysource.com/issues/43618054-question-http-test-to-bypass-insecure-https-url).

### CODE REVIEW

File: src/main/java/nl/hsac/fitnesse/fixture/Environment.java [hunk 1]
```diff
@@ -781,4 +781,14 @@ public void enableHttpClientCompression() {
     public void disableHttpClientCompression() {
         httpClient.disableCompression();
     }
+
+    /**
+     * Disables SSL certificate verification on this environment's HttpClient
+     */
+    public void disableHttpClientSSLVerification() { httpClient.disableSSLVerification(); }
```
Comment (fhoeben) [inline src/main/java/nl/hsac/fitnesse/fixture/Environment.java]: Please split over 3 lines.

File: src/main/java/nl/hsac/fitnesse/fixture/Environment.java [hunk 2]
```diff
@@ -781,4 +781,14 @@ public void enableHttpClientCompression() {
...
+
+    /**
+     * Enables SSL certificate verification on this environment's HttpClient
+     */
+    public void enableHttpClientSSLVerification() { httpClient.enableSSLVerification(); }
```
Comment (fhoeben) [inline src/main/java/nl/hsac/fitnesse/fixture/Environment.java]: Please split over 3 lines.

File: src/main/java/nl/hsac/fitnesse/fixture/slim/HttpTest.java [hunk 3]
```diff
@@ -66,6 +66,16 @@ public void disableCompression() {
         getEnvironment().disableHttpClientCompression();
     }
 
+    /**
+     * Disables SSL certificate verification in the current environment (i.e. for the entire test run)
+     */
+    public void disableSSLVerification() { getEnvironment().disableHttpClientSSLVerification(); }
```
Comment (fhoeben) [inline src/main/java/nl/hsac/fitnesse/fixture/slim/HttpTest.java]: Please split over 2 lines and properly indent closing brace

File: src/main/java/nl/hsac/fitnesse/fixture/slim/HttpTest.java [hunk 4]
```diff
@@ -66,6 +66,16 @@ public void disableCompression() {
...
+
+    /**
+     * Enables SSL certificate verification in the current environment (i.e. for the entire test run)
+     */
+    public void enableSSLVerification() { getEnvironment().enableHttpClientSSLVerification(); }
```
Comment (fhoeben) [inline src/main/java/nl/hsac/fitnesse/fixture/slim/HttpTest.java]: Please split over 3 lines.

File: src/main/java/nl/hsac/fitnesse/fixture/util/HttpClient.java [hunk 5]
```diff
@@ -39,11 +42,11 @@
 public class HttpClient {
     /** Default HttpClient instance used. */
     public final static org.apache.http.client.HttpClient DEFAULT_HTTP_CLIENT;
-    public static org.apache.http.client.HttpClient COMPRESSION_HTTP_CLIENT;
     private org.apache.http.client.HttpClient httpClient;
+    private boolean contentCompression = false, sslVerification = true;
```
Comment (fhoeben) [inline src/main/java/nl/hsac/fitnesse/fixture/util/HttpClient.java]: Please put sslVerification field on own line

File: wiki/FitNesseRoot/HsacExamples/SlimTests/HttpTests/HttpsWithoutCertificateVerification.wiki [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+'''''Use at your own risk, this makes https potentially as unsafe as http!'''''
+
+|script  |http test                         |
+|disable SSL verification                   |
+|get from|https://expired.badssl.com/       |
+|get from|https://wrong.host.badssl.com/    |
+|get from|https://self-signed.badssl.com/   |
+|get from|https://untrusted-root.badssl.com/|
+
+Note that 'disable SSL verification' lasts for the rest of the test run, or until you call 'enable SSL verification'.
```
Comment (fhoeben) [inline wiki/FitNesseRoot/HsacExamples/SlimTests/HttpTests/HttpsWithoutCertificateVerification.wiki]: Please add a new table at bottom of page, to enable SSL verification again

**Comment by Fastigium** [comment]
Requested changes made! My IDE sneakily auto-folded the surrounding methods, I'll have to watch for that in the future.

**Comment by fhoeben** [comment]
Thanks!

**Comment by GowthamX** [comment]
Hi, 
How can i pass certificates to my Http test.
I tried defining the path in Wiki file and also in command prompt while opening the fitnesse jar.
Please help

**Comment by tcnh** [comment]
Have a look at: https://github.com/fhoeben/hsac-fitnesse-fixtures/issues/123
I think this thread adresses your question. (In short: create a keystore containing your cert and pass it to slim using the command pattern)

**Comment by GowthamX** [comment]
Caused by: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target

Hi tried the Command Pattern as mentioned, still no luck

!define COMMAND_PATTERN {java -Djavax.net.ssl.truststore=C:\Users\gowth\Documents\Fitnesse\hsac-fitnesse-fixtures-snapshot-standalone\1532355493751\client.jks -Djavax.net.ssl.truststore=JKS -Djavax.net.ssl.truststorePassword=test -cp %p %m}

**Pull request merged**